skypy/halos/sham.py

```diff
@@ -0,0 +1,1234 @@
+
+            elif ID_A > 0 and bs_counter != len(bs_order):  # Subhalo assigned blue satellite
+                gal_assigned.append(bs_order[bs_counter])
+                gal_type_A.append(4)
+                bs_counter += 1
+
+            else:  # No blue to assign
+                del_ele.append(ii)
+
+    # Delete and array final lists
+    hs_fin = np.delete(hs_order, del_ele)
+    id_fin = np.delete(id_order, del_ele)
+    z_fin = np.delete(z_order, del_ele)
+    gal_fin = np.array(gal_assigned)
+    gal_type_fin = np.array(gal_type_A)
+
+    return hs_fin, gal_fin, id_fin, z_fin, gal_type_fin
+
+
+def scatter_proxy(gal):
+    r'''Function to add scatter to galaxies using a proxy mass
+
+    This function takes the generated galaxy masses from the catalogues
+    and generates a gaussian proxy mass for each, then reorders based on
+    this proxy mass. This adds scatter to the galaxies.
+
+    Parameters
+    -----------
+    gal : (nm, ) array_like
+        List of galaxy masses to be scattered, in units of solar mass
+
+    Returns
+    --------
+    new_gal : (nm, ) array_like
+        Reordered galaxy masses according to generated proxy mass, highest
+        to lowest mass, in units of solar mass
+
+    Examples
+    ---------
+    >>> from skypy.halos import sham
+
+    This example generates a galaxy catalogue and then reorders it
+
+    >>> gal_cat = sham.run_file('galaxies.yaml', galaxy, mass)
+    >>> new_gal = sham.scatter_proxy(gal_cat)
+
+    References
+    ----------
+    ..
+    '''
+    # Errors
+    gal = np.atleast_1d(gal)
+    if ((gal <= 0)).any():
+        raise Exception('Galaxy masses must be positive and non-zero')
+
+    # Draw one sample for each galaxy
+    # Use galaxy mass as mean and sigma of one order less
+    proxy_gal = np.random.normal(loc=gal, scale=gal/10)
+
+    # Order by proxy mass
+    proxy_stack = np.stack((proxy_gal, gal), axis=1)
+    gal_stack = proxy_stack[np.argsort(proxy_stack[:, 0])]
+    return np.flip(gal_stack[:, 1])  # Return new galaxies, highest mass first
+
+
+# Separate populations
+def sham_plots(hs_fin, gal_fin, gal_type_fin, print_out=False):
+    r'''Function that pulls assigned galaxies together into groups for
+    plotting
+
+    This function takes in the final assigned halos and galaxies and splits
+    them into each type of galaxy for plotting and comparing to data. It also
+    computes the separate central and satellite populations with no colour
+    split.
+
+    Parameters
+    -----------
+    hs_fin : (nm, ) array_like
+        List of assigned halos and subhalos, in units of solar mass. The
+        subhalos are assumed to be their current stripped mass
+    gal_fin : (nm, ) array_like
+        List of assigned galaxies, in units of solar mass
+    gal_type_fin : (nm, ) array_like
+        List of assigned galaxies types with the tags 1,2,3,4 for
+        red centrals, red satellites, blue centrals, blue satellites
+    print_out : Boolean

```

Comment:
The skypy pipeline call includes a "verbose" flag which could be used instead of this `print_out` keyword.

Reply:
Thanks for the review. I am struggling to work out how to interact with the verbose keyword as I can't see that it is used anywhere in the galaxies module as an example. From the documentation, this would be applied to the code as a command line argument and I wouldn't expect this code to be used as such due to the dictionary output. Any help would be appreciated.